SFS_general/sphbesselh_zeros.m

```diff
@@ -53,74 +66,114 @@
+z = 1./r;
+
+end
+
+
+%% ===== Auxiliary Functions =============================================
+function z0 = campos_zeros(n)
+    % Approximate roots of ordinary Bessel polynomial of nth order, see
+    % Campos and Calderon (2011)
+
+    if n == 1
+        z0 = -1;
+        return
+    end
+
+    % See CamposCalderon2011, Eq. (7)
+    r = polyval( [1 2 0 0],n);          % n^3 + n^2
+    a1 = polyval( [-6 -6],n) / r;       % ( -6n - 6 ) / r
+    a2 = 6 / r;
+
+    s = polyval( [1 -3 0 2 0 0],n);     % n^5 - 3n^4 + 2n^2
+    b3 = polyval( [-8 16],n) / s;       % ( -8n + 16 ) / s
+    b2 = polyval( [12 -12 -24],n) / s;  % ( 12n^2 - 12n - 24 ) / s
+    b1 = polyval( [-2 -12 24 8],n) / s; % ( -2n^3 - 12n^2 + 24n + 8 ) / s
+    b0 = polyval( [-1 5 0 -6 0],n) / s; % ( -n^4 + 5n^3 - 6n ) / s
+
+    % See Campos and Calderon (2011), Eq. (4)
+    k = (1:n).';
+    x0 = polyval( [a2 a1 0],k);         % real part
+    y0 = polyval( [b3 b2 b1 b0],k);     % imaginary part
+
+    % See Campos and Calderon (2011), Eq. (8)
+    z0 = x0 + 1j*y0;
+
+end
+
+function x = aberth(f,fp,x0,TOL,MAXITER)
+    % Ehrlich-Aberth method to simulatenous approximation of roots
+    N = length(x0);
+    beta = zeros(size(x0));
+    x = x0;
+    for iter=1:MAXITER
+        alpha = -f(x) ./ fp(x);
+        
+        for k=1:N
+            beta(k) = sum(1./(x(k) - x([1:k-1,k+1:N])));
         end
-        for n=0:order
-          % Flip nominator polynoms
-          B{n+1} = B{n+1}(end:-1:1);
-          % Zeros
-          z{n+1} = roots(B{n+1});
-          % Poles (are always zero)
-          p{n+1} = zeros(order,1);
+        
+        x = x + alpha./(1 + alpha .* beta);
+        
+        if all(abs(alpha) <= TOL)
+            return
         end
-        z = z{order+1};
-        p = p{order+1};
-    else
-        error('%s: method has to be 1 or 2.',upper(mfilename));
     end
-else
-    error(['%s: for orders higher than 85 no stable numerical ', ...
-           'method is available at the moment to caclulate the zeros.'], ...
-          upper(mfilename));
+    warning('%s: maximum iterations in aberth() reached; order = %d', ... 
+        upper(mfilename), N);
 end
-return
-
-
-%% ===== Computation with Multiprecission Toolbox ========================
-% For the Multiprecission Toolbox, see: http://www.advanpix.com
-% Unfortunately it turned out, that the obtained zeros with this method have
-% some systematic errors, see
-% https://github.com/sfstoolbox/sfs/issues/57#issuecomment-183791477
-% The following code was used to calculate the zeros with the Multiprecission
-% Toolbox. The results are stored at
-% https://github.com/sfstoolbox/data/tree/master/sphbesselh_zeros
-B = mp(zeros(1,order+2));
-A = B;
-for n=mp(0:order)
-    B(n+1) = factorial(2*order-n)/(factorial(order-n)*factorial(n)*2^(order-n));
+
+function x = newton(f,fp,x0,TOL,MAXITER)
+    % Newton-Rapheson method for approximation of a single root
```

Comment:
The scipy implementation does handle many special cases and has some extensions. Here, the most simple version of the Newton-Rapheson method is implemented.